hazelcast/src/main/java/com/hazelcast/map/impl/record/DataRecordFactory.java

```diff
@@ -65,5 +72,35 @@ public void setValue(Record<Data> record, Object value) {
             v = serializationService.toData(value, partitionStrategy);
         }
         record.setValue(v);
+        updateValueMetadata(record, v);
+    }
+
+    private void updateValueMetadata(Record<Data> record, Data dataValue) {
+        try {
+            Object valueMetadata = metadataInitializer.createFromData(dataValue);
```

Comment:
Is it more performant if the parser works on top of binary or String values?

Reply:
Parser is more performant on string values than binary. However, the cost of deserializing the whole string before feeding it to the parser reverses the benefit easily.

Reply:
Just thinking, we can pass in `value` directly from `setValue` and use the deserialized instance, if possible. If the `value` is `Data`, we fallback to parsing the binary value.

Reply:
If I understand correctly, you are suggesting creating metadata from object form if available. This does not work because metadata has to be created from the queried form. Since this is DataRecordFactory we want to query binary form. So we have to create the metadata from binary form.

Reply:
Yes, I was thinking something like this:
```java
    @Override
    public void setValue(Record<Data> record, Object value) {
        assert value != null : "value can not be null";

        final Data v;
        if (value instanceof Data) {
            v = (Data) value;
        } else {
            v = serializationService.toData(value, partitionStrategy);
        }
        record.setValue(v);
        updateValueMetadata(record, value);
    }

    private void updateValueMetadata(Record<Data> record, Object dataValue) {
        try {
            Object valueMetadata = dataValue instanceof Data 
                    ? metadataInitializer.createFromData((Data) dataValue) 
                    : metadataInitializer.createFromObject(dataValue);
            Metadata existing = record.getMetadata();
            if (existing != null) {
                existing.set(false, valueMetadata);
            }
        } catch (Exception e) {
            rethrow(e);
        }
    }
```
If we must create the metadata from the binary format, ignore the suggestion.